keps/sig-node/3317-user-namespaces/README.md

```diff
@@ -0,0 +1,1015 @@
+proposed in this KEP.
+
+- Old runtime: idem, but CRI proto files don't include this changes.
+
+New runtime and old kubelet: all works just fine. Kubelet doesn't request userns
+(doesn't have that feature) and therefore the runtime doesn't create them. The
+runtime can detect this situation as the `user` field in the `NamespaceOption`
+will be seen as nil, [thanks to protobuf][proto3-defaults]. We already tested
+this with real code.
+
+Old runtime and new kubelet: all will work fine. As the `user` field of the
+`NamespaceOption` message is not part of the runtime protofiles, that part is
+ignored by the runtime and pods are created using the host userns.
+
+
+[proto3-defaults]: https://developers.google.com/protocol-buffers/docs/proto3#default
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: UserNamespacesSupport
+  - Components depending on the feature gate: kubelet
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Pods that were created and opted-in to use user namespaces, will have to be
+recreated to run again without user namespaces.
```

Comment:
Can the pods be safely deleted (without leaking anything) after the feature was disabled?

Or should we ensure that those pods are deleted *before* disabling the feature gate?

Reply:
Good question.

Yes, pods can be safely deleted even after the feature was disabled in the kubelet. We just store the mapping for this pod in /var/lib/kubelet/$POD/userns, the pod deletion takes care of deleting everything in that directory so we don't leak anything.

I've just tested it locally, to be sure I was not missing anything :)